### ISSUE TITLE
Quantization loss wrapper?

### ISSUE THREAD

**Issue by h4gen** [UKPLab/sentence-transformers#2829]
Hi 👋 

I am wondering if it makes sense to write a wrapper for losses like `MatryoshkaLoss` in order to incorporate quantization into the model training, i.e. a `BinaryQunatizationLoss` that takes another loss as argument. I am mainly wondering because I would assume that this should definitely improve the retrieval quality of the model if we already know during training that quantization will be applied afterwards, analogous to the Matryoshka technique. Having said this, I find virtually no information or research that this is meaningful or that somebody has done that already. 

Any thoughts on this?

Best,
Hagen

**Comment by tomaarsen** [comment]
Hello!

The primary bottleneck for not having a "quantization" loss right now is the lack of serious research on this matter. The only paper that I know of is BPR ([Yamada et al., 2021](https://arxiv.org/abs/2106.00882)), but my tests with [bpr_loss.py](https://github.com/beir-cellar/beir/blob/main/beir/losses/bpr_loss.py) did not produce very useful results: just training with MultipleNegativesRankingLoss gave me better results.

In short: 
1. No, there isn't currently a loss "modifier" like MatryoshkaLoss that allows for training for quantization.
2. Yes, I would be very interested in one.

Now, I think implementing a loss modifier like MatryoshkaLoss is sadly infeasible: Although it's possible to create a loss function that modifies the embedding model to produce quantized embeddings & then call some other loss, that other loss will possibly not respond well to these new unexpected quantized embeddings. For example, if we wrapped MultipleNegativesRankingLoss, then we would probably have to also update the similarity function [here](https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/losses/MultipleNegativesRankingLoss.py#L97).

That said, I'm also open to e.g. updating all losses to rely on [model.similarity](https://sbert.net/docs/package_reference/sentence_transformer/SentenceTransformer.html#sentence_transformers.SentenceTransformer.similarity), and then add e.g. the "hamming" distance metric (useful for binary). That way, a (binary) quantization loss modifier might make sense & could work with most losses out of the box.

- Tom Aarsen

**Comment by zxzl** [comment]
@h4gen I had similar intest and you can refer to https://arxiv.org/pdf/2302.08714, but it's not directly applicable to sentencetransformers. But hope you can get some inspirations here

**Comment by h4gen** [comment]
Hey @tomaarsen and @zxzl,

Thanks to both of you for your responses. @tomaarsen What you say makes total sense. Would you do it such that `precision` is an attribute of the loss and that hamming is used in case the precision is binary or would `precision` become an attribute of the model itself and the loss behaves respectively? I will try out the approach next week and share some "anecdotal evidence".

@zxzl That looks super interesting and impressive on first glance and very promising results.  Will do a deeper read the next days, thanks for sharing! Do you have an opinion how well a method such as proposed by @tomaarsen would work or where you would see the lowest hanging fruits building on top of/in sentencetransformers?

Best,
Hagen

**Comment by zxzl** [comment]
@h4gen Glad that you found the paper interesting. 

>That said, I'm also open to e.g. updating all losses to rely on [model.similarity](https://sbert.net/docs/package_reference/sentence_transformer/SentenceTransformer.html#sentence_transformers.SentenceTransformer.similarity), and then add e.g. the "hamming" distance metric (useful for binary). That way, a (binary) quantization loss modifier might make sense & could work with most losses out of the box.

I think you can try this on your data and compare how performance changes. At least model should work better after quantized.

**Comment by h4gen** [comment]
Okay I tried it out and find in surprisingly underwhelming so far. Inspired by [this article about STE](https://hassanaskary.medium.com/intuitive-explanation-of-straight-through-estimators-with-pytorch-implementation-71d99d25d9d0) which @zxzl also mentions in the paper I came up with the following similarity function:

```
class STEFunction(torch.autograd.Function):
    @staticmethod
    def forward(ctx, input):
        return (input >= 0).float()

    @staticmethod
    def backward(ctx, grad_output):
        return grad_output

class StraightThroughEstimator(nn.Module):
    def __init__(self):
        super(StraightThroughEstimator, self).__init__()

    def forward(self, x):
        return STEFunction.apply(x)

def _convert_to_tensor(x):
    if isinstance(x, list):
        x = torch.tensor(x)
    elif isinstance(x, np.ndarray):
        x = torch.from_numpy(x)
    return x.float()

def pairwise_binary_hamming_similarity(a: list | np.ndarray | torch.Tensor, b: list | np.ndarray | torch.Tensor) -> torch.Tensor:

    a = _convert_to_tensor(a)
    b = _convert_to_tensor(b)

    ste = StraightThroughEstimator()

    bin_a = ste(a)
    bin_b = ste(b)

    # Expand dimensions to compute pairwise differences
    bin_a_expanded = bin_a.unsqueeze(1)  # shape (m, 1, d)
    bin_b_expanded = bin_b.unsqueeze(0)  # shape (1, n, d)

    pairwise_hamming_distances = torch.abs(bin_a_expanded - bin_b_expanded).mean(dim=-1)  # shape (m, n)

    pairwise_normalized_similarity = 1 - pairwise_hamming_distances

    return pairwise_normalized_similarity
```

I use it as drop-in replacement for `similarity_fct` like for `MultipleNegativesSymmetricRankingLoss` to not touch the ST codebase too much for now. It works but tldr from my anecdotal evidence so far is that it does not perform better. In general there are two findings:

1. Using this approach yields a higher score for binary hamming sim than all other metrics relatively
2. It doesn't matter as just training w/o and binarize the embeddings afterwards will still yield better results in absolute terms

Going to pt this aside for now as I have more pressing stuff on my plate but maybe this sparks something for someone.

Best,
Hagen
